### PR TITLE
feat(permissions): filter fields in forms based on read/write permissions

### PR DESCRIPTION
**Pull request by tomwwinter**
- filter fields in forms with no read permission
- disable fields in edit mode with read-only permissions

Can be tested with this permissions:

```js
const rules: DatabaseRules = {
  user_app: [
    { subject: "Child", action: "read", fields: ["name"] },
    { subject: "Child", action: "manage", fields: ["name", "dateOfBirth"] },
  ],
  admin_app: [{ subject: "all", action: "manage" }],
}
```

related to: #1912

### Visible/Frontend Changes
- [x] filter fields with now read permission in detail view 
- [x] disable fields with no write access in edit mode 

### Architectural/Backend Changes
- [ ] none

### CODE REVIEW

File: src/app/core/entity-details/form/form.component.ts [hunk 1]
```diff
@@ -32,28 +34,49 @@ import { FieldGroup } from "./field-group";
...
     private router: Router,
     private location: Location,
     private entityFormService: EntityFormService,
     private alertService: AlertService,
+    private abilityService: AbilityService,
   ) {}
 
+  ngOnDestroy(): void {
+    this._destroyed.next();
+    this._destroyed.complete();
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/form/form.component.ts:61]: We use the [@ngneat/until-destroy](https://www.npmjs.com/package/@ngneat/until-destroy) package to make the unsubscribing a little easier (e.g. see the `NoteDetailsComponent` on how to use it)

File: src/app/core/entity-details/form/form.component.ts [hunk 2]
```diff
@@ -82,6 +105,25 @@ export class FormComponent<E extends Entity> implements FormConfig, OnInit {
     this.entityFormService.resetForm(this.form, this.entity);
     this.form.disable();
   }
+
+  private filterFieldGroupsByPermissions() {
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/form/form.component.ts:109]: Do you see any way on how we can move (most of) this logic into the `EntityFormService` or the `EntityFormComponent` so we can re-use the functionality in all forms across the map?

File: src/app/core/entity-details/form/form.component.ts [hunk 3]
```diff
@@ -82,6 +105,25 @@ export class FormComponent<E extends Entity> implements FormConfig, OnInit {
...
+          this.abilityService.hasReadPermission(this.entity, field),
+        );
+        return group;
+      })
+      .filter((group) => group.fields.length !== 0);
+  }
+
+  private disableReadOnlyFormControls() {
+    Object.keys(this.form.controls).forEach((fieldId) => {
+      if (!this.abilityService.hasEditPermission(this.entity, fieldId)) {
```
Comment (TheSlimvReal) [inline src/app/core/entity-details/form/form.component.ts:122]:
```suggestion
      if (!this.ability.can('read', this.entity, fieldId)) {
```
This way you wouldn't event need the extra functions in the ability service.

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
Thanks for the goods start already. The functionality looks good, lets try to refactor this so we can use the same approach in all forms.

File: src/app/core/entity-details/form/form.component.ts (hunk 1 above)
Comment (tomwwinter) [reply src/app/core/entity-details/form/form.component.ts]: done

File: src/app/core/permissions/ability/ability.service.ts [hunk 4]
```diff
@@ -27,6 +39,14 @@ export class AbilityService extends LatestEntityLoader<Config<DatabaseRules>> {
     super(Config, Config.PERMISSION_KEY, entityMapper, logger);
   }
 
+  hasReadPermission(entity: Entity, field: string | FormFieldConfig): boolean {
+    return this._hasFieldPermission(entity, field, this.READ_PERMISSION_GROUP);
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/permissions/ability/ability.service.ts]: done.

File: src/app/core/entity-details/form/form.component.ts (hunk 2 above)
Comment (tomwwinter) [reply src/app/core/entity-details/form/form.component.ts]: Logic moved to EntityFormComponent and EntityFormService

File: src/app/core/entity-details/form/form.component.ts (hunk 3 above)
Comment (tomwwinter) [reply src/app/core/entity-details/form/form.component.ts]: done

File: src/app/core/common-components/entity-form/entity-form/entity-form.component.ts [hunk 5]
```diff
@@ -78,6 +81,31 @@ export class EntityFormComponent<T extends Entity = Entity>
...
+  private applyFormFieldPermissions() {
+    if (this.fieldGroups) {
+      this.fieldGroups = this.entityFormService.filterFieldGroupsByPermissions(
+        this.fieldGroups,
+        this.entity,
+      );
+    }
+
+    if (this.form) {
+      this.form.statusChanges.pipe(untilDestroyed(this)).subscribe((_) => {
```
Comment (TheSlimvReal) [inline src/app/core/common-components/entity-form/entity-form/entity-form.component.ts:102]: I guess this does not need to be run for every status change (e.g. not when the status changes to `disabled`).

File: src/app/core/common-components/entity-select/entity-select.component.spec.ts [hunk 6]
```diff
@@ -36,6 +38,8 @@ describe("EntitySelectComponent", () => {
         ]),
       ],
     }).compileComponents();
+    entityAbility = TestBed.inject(EntityAbility);
+    spyOn(entityAbility, "can").and.returnValue(true);
```
Comment (TheSlimvReal) [inline src/app/core/common-components/entity-select/entity-select.component.spec.ts:42]: Thats a bit annoying, maybe we should add the default "everything is allowed"-rule to the `MockedTestingModule`.

File: src/app/core/common-components/entity-form/entity-form.service.ts [hunk 7]
```diff
@@ -78,6 +79,35 @@ export class EntityFormService {
     }
   }
 
+  filterFieldGroupsByPermissions<T extends Entity = Entity>(
+    fieldGroups: FieldGroup[],
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form.service.ts:83]: It's a lot of logic for be handled inside the component, so i would keep inside the service. But lets discuss this in a call.

File: src/app/core/common-components/entity-form/entity-form/entity-form.component.ts (hunk 5 above)
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form/entity-form.component.ts:102]: I will add a check for disabled state.

File: src/app/core/common-components/entity-select/entity-select.component.spec.ts (hunk 6 above)
Comment (tomwwinter) [reply src/app/core/common-components/entity-select/entity-select.component.spec.ts:42]: Problem is also that our "unit tests" are more like integration tests with a lot of real dependencies... But yes, the default could be to allow everything.

File: src/app/core/common-components/entity-form/entity-form/entity-form.component.ts [hunk 8]
```diff
@@ -78,6 +81,31 @@ export class EntityFormComponent<T extends Entity = Entity>
...
+        this.entityFormService.disableReadOnlyFormControls(
+          this.form,
+          this.entity,
+        );
+      });
+    }
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form/entity-form.component.ts]: Implemented

File: src/app/core/common-components/entity-form/entity-form.service.spec.ts [hunk 9]
```diff
@@ -54,6 +55,46 @@ describe("EntityFormService", () => {
...
+    TestBed.inject(EntityAbility).update([
+      { subject: "Entity", action: "update", fields: ["foo"] },
+    ]);
+
+    service.disableReadOnlyFormControls(formGroup, entity);
+
+    expect(formGroup.get("name").disabled).toBeTruthy();
+    expect(formGroup.get("foo").disabled).toBeFalse();
+    expect(formGroup.get("bar").disabled).toBeTruthy();
+  });
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form.service.spec.ts:73]: Test created

File: src/app/core/common-components/entity-form/entity-form/entity-form.component.ts [hunk 10]
```diff
@@ -78,6 +83,22 @@ export class EntityFormComponent<T extends Entity = Entity>
       this.initialFormValues = this.form.getRawValue();
       this.disableForLockedEntity();
     }
+
+    if (this.form && this.entity) {
+      this.entityFormService.disableReadOnlyFormControls(
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form/entity-form.component.ts:88]: Good point, i forgot to remove it, after moved to the service ;)

File: src/app/core/common-components/entity-form/entity-form/entity-form.component.ts [hunk 11]
```diff
@@ -78,6 +83,22 @@ export class EntityFormComponent<T extends Entity = Entity>
...
+        this.form,
+        this.entity,
+      );
+    }
+  }
+
+  private applyFormFieldPermissions() {
```
[the remark this answers is not in the document]
Comment (tomwwinter) [reply src/app/core/common-components/entity-form/entity-form/entity-form.component.ts:95]: The function was quite long and therefore easier to read. However, as it has now become smaller, I have dragged it back into the function.

**Review by TheSlimvReal — COMMENTED** [review]
Code looks very good now. I only changed some small things. When playing with it I discovered a functionality issue though:

Currently the `disabled` state is always applied based on the `update` permissions. This means that the property is also disabled when a new entity (e.g. a new Note) is created even if the user has `create` permissions on the property. I think there is a valid use case where users are allowed to set a property initially, but once the entity already exists they are not allowed to change it anymore. Therefore I think for new entities the disabled state should be evaluated based on the `create` permission for a field.

**Review by TheSlimvReal — APPROVED** [review]
Works great now. Thanks.

**Pull request merged**